### PR TITLE
Create a mock backend service so API calls from Firebase can be tested in isolation

### PR DESCRIPTION
**Pull request by Mapleia**
## Changes
Create a mock express app that can be used for integration tests. This results in two distinct test workflows: One using the mocked API (which should be updated by the companion-app developer) and one referencing the latest version of the `data-services` API pulled from the `master` branch. The latter will fail independently in the case that the backend repo is not up to date.

Requires: 
- `yarn install` in the `mock/api` directory
    - express, cors, typescript
- node version >=14

## PR Dependencies
None.

## Subtasks / Progress
- [x] Debug export test functions
- [x] Add spinning up the mock api server in the CI workflow

## Testing
- [x] Automated

### CODE REVIEW

File: common/controllers/StorageController.ts [hunk 1]
```diff
@@ -1,8 +1,10 @@
-import FileMetadata = firebase.storage.UploadMetadata;
+import firebase from 'firebase';
 import Firebase from 'common/services/firebase';
 import { createLogger } from 'common/logger';
 import Lazy from 'common/utils/lazy';
 
+type FileMetadata = firebase.storage.UploadMetadata;
+
```
Comment (rlnsy) [inline common/controllers/StorageController.ts]: what's going on with this change here?

File: common/database/repositories/ClientCardRepo.ts [hunk 2]
```diff
@@ -35,7 +35,6 @@ class ClientCardRepo {
         const qu = query.where('email', '==', email);
 
         const snapshot = await qu.get();
-
```
Comment (rlnsy) [inline common/database/repositories/ClientCardRepo.ts]: Let's try to avoid unrelated whitespace changes wherever possible

File: server/functions/__tests__/export.spec.ts [hunk 3]
```diff
@@ -2,11 +2,17 @@ import * as admin from 'firebase-admin';
 import { ExportFunctions } from '../src/export';
 import { init } from './util/firebase';
 import { assert } from 'chai';
-
+import Collections from 'common/database/collections';
+// import { initializeAsync } from '../../../common/services/firebase';
+// import clientConfig from './mocks/client/config';
```
Comment (rlnsy) [inline server/functions/__tests__/export.spec.ts:7]: let's remove these

File: server/functions/__tests__/mocks/api/index.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    } else if (isNaN(data.data.date)) {
+        res.status(400).send({
+            message: 'date must be a number',
+        });
+    } else {
+        res.status(201).send({
+            message: 'Created',
+        });
+    }
+});
```
Comment (rlnsy) [inline server/functions/__tests__/mocks/api/index.ts:50]: would it be possible to add a basic state variable (key-value map is fine) and an endpoint to get the state? We can use that to test which data is sent to the API.

**Review by rlnsy — CHANGES_REQUESTED** [review]
Some comments below. Also, major thing: I was hoping that the mock api could live alongside the actual api in the test workflow. Can you set up two separate jobs in the GH workflow ,one using the old API config and one using this? This way we can see differences between the actual API and the mock.

File: common/controllers/StorageController.ts (hunk 1 above)
Comment (Mapleia) [reply common/controllers/StorageController.ts]: I think that's a change from the export-functions. I've compared export-functions vs mock-backend, and it's not a change from the mock-backend branch.

Comment (Mapleia) [reply common/controllers/StorageController.ts]: You can confirm it with the compare branch tool from `GitLens` extension in VSCode.

File: common/database/repositories/ClientCardRepo.ts (hunk 2 above)
Comment (Mapleia) [reply common/database/repositories/ClientCardRepo.ts]: Got it! I'm not seeing that files in the list of changed files though?

File: server/functions/__tests__/export.spec.ts (hunk 3 above)
Comment (Mapleia) [reply server/functions/__tests__/export.spec.ts:7]: Removed!

File: server/functions/__tests__/mocks/api/index.ts (hunk 4 above)
Comment (Mapleia) [reply server/functions/__tests__/mocks/api/index.ts:82]: Yep it's possible! Added as 
```
data: {
    uid: string,
    value: number,
    name: string,
    source: string,
    date: number,
}
```

**Comment by Mapleia** [comment]
> Some comments below. Also, major thing: I was hoping that the mock api could live alongside the actual api in the test workflow. Can you set up two separate jobs in the GH workflow ,one using the old API config and one using this? This way we can see differences between the actual API and the mock.

The original test was added back!

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -12,7 +27,7 @@ jobs:
         uses: actions/checkout@v2
         with:
           repository: bipolarbridges/data-services
-          ref: export-api-v0
+          ref: qol-survey
```
Comment (rlnsy) [inline .github/workflows/tests.yml:30]: We should change this to `master` now. This test will fail, but it will provide us an indication of the state of integration with the actual data-services. The mock version will provide a more accurate and up-to-date version of the tests results. (Plus, we can change the ref for development purposes in subsequent PRs, but it should default to `master`.

File: docker-compose.yml [hunk 6]
```diff
@@ -1,5 +1,10 @@
 version: '3'
 services:
+  api:
```
Comment (rlnsy) [inline docker-compose.yml:3]: let's use `mock-api` or smth

File: server/functions/__tests__/mocks/api/.nvmrc [hunk 7]
```diff
@@ -0,0 +1 @@
+14
```
Comment (rlnsy) [inline server/functions/__tests__/mocks/api/.nvmrc:1]: Should mention this in the requirements section of the PR.

File: server/functions/src/export.ts [hunk 8]
```diff
@@ -86,12 +86,12 @@ fns.measurement = FeatureSettings.ExportToDataServices
             const data: RecordData = snap.data() as RecordData;
             const backend = new FunctionBackendController();
             const makeRequest = async (ex: RecordExport) =>
-                backend.logMeasurement(data.clientUid, data.coachUid, ex.typeId, ex.value, data.date)
+                backend.logMeasurement(data.clientUid, 'malso-measurement', ex.typeId, ex.value, data.date)
```
Comment (rlnsy) [inline server/functions/src/export.ts:89]: typo in 'maslo'

**Review by rlnsy — CHANGES_REQUESTED** [review]
Looks nice just a few nitpicks

File: docker-compose.yml (hunk 6 above)
Comment (Mapleia) [reply docker-compose.yml:3]: We need to share the name of the mock & not mocked api so github can connect to both. The .env variable for BACKEND_ADDRESS is set to =api, and we can't really change that during run time?

File: server/functions/__tests__/mocks/api/.nvmrc (hunk 7 above)
Comment (Mapleia) [reply server/functions/__tests__/mocks/api/.nvmrc:1]: Added.

File: server/functions/src/export.ts (hunk 8 above)
Comment (Mapleia) [reply server/functions/src/export.ts]: Fixed.

File: .github/workflows/tests.yml (hunk 5 above)
Comment (Mapleia) [reply .github/workflows/tests.yml:30]: I see! Make sense given that the one that would be deployed in production would be the master version.

File: docker-compose.yml (hunk 6 above)
Comment (rlnsy) [reply docker-compose.yml:3]: Excellent point! I forgot about the docker-based networking. Fine to leave as is for now.

**Review by rlnsy — APPROVED** [review]

**Comment by rlnsy** [comment]
hallelujah

**Comment by wynnset** [comment]
@Mapleia seems like our server tests are failing

**Review by wynnset — APPROVED** [review]
Code looks good. Can you please just add detailed instructions for using the mock backend server into the README and/or on Notion?

**Comment by Mapleia** [comment]
> Code looks good. Can you please just add detailed instructions for using the mock backend server into the README and/or on Notion?

For sure! I'll have it done within the next hour. 
I'll also look into that server test fail it seems like the input doesn't match the spec anymore?

**Comment by Mapleia** [comment]
@wynnset The test with our actual data-service is expected to fail when master branch doesn't contain the changes that is required for the export functions. However, with the recent merged on data-service, it's okay for now. The tests we should really be worried about are the mock-backend ones because that points to a larger issue.

**Comment by Mapleia** [comment]
Added to the Notion and readme!
